token-erc-721/chaincode-javascript/lib/tokenERC721.js

```diff
@@ -0,0 +1,484 @@
+    }
+
+
+    // ============== Extended Functions for this sample ===============
+
+    /**
+     * Set optional information for a token.
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} name The name of the token
+     * @param {String} symbol The symbol of the token
+     */
+    async SetOption(ctx, name, symbol) {
+        await ctx.stub.putState(nameKey, Buffer.from(name));
+        await ctx.stub.putState(symbolKey, Buffer.from(symbol));
+        return true;
+    }
+
+    /**
+     * Mint a new non-fungible token
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} tokenId Unique ID of the non-fungible token to be minted
+     * @returns {Object} Return the non-fungible token object
+    */
+    async Mint(ctx, tokenId) {
+
+        // Check minter authorization - this sample assumes Org1 is the issuer with privilege to mint a new token
+        const clientMSPID = ctx.clientIdentity.getMSPID();
+        if (clientMSPID !== 'Org1MSP') {
+            throw new Error('client is not authorized to mint new tokens');
+        }
+
+        // Get ID of submitting client identity
+        const minter = ctx.clientIdentity.getID();
+
+        // Check if the token to be minted does not exist
+        const exists = await this._nftExists(ctx, tokenId);
+        if (exists) {
+            throw new Error(`The token ${tokenId} is already minted.`);
+        }
+
+        // Add a non-fungible token
+        const tokenIdInt = parseInt(tokenId);
+        const nft = { tokenId: tokenIdInt, owner: minter };
+        const nftKey = ctx.stub.createCompositeKey(nftPrefix, [tokenId]);
+        await ctx.stub.putState(nftKey, Buffer.from(JSON.stringify(nft)));
+
+        // Update balance
+        await this._addBalance(ctx, minter, tokenId);
+
+        // Add the tokenId to the global token array for index management
+        const tokensAsBytes = await ctx.stub.getState(tokenArrayKey);
+        let tokens;
+        if (!tokensAsBytes || tokensAsBytes.length === 0) {
+            tokens = [];
+        } else {
+            tokens = JSON.parse(tokensAsBytes.toString());
+        }
+        tokens.push(tokenIdInt);
+        await ctx.stub.putState(tokenArrayKey, Buffer.from(JSON.stringify(tokens)));
+
+        // Emit the Transfer event
+        const transferEvent = { from: '0x0', to: minter, tokenId: tokenIdInt };
+        ctx.stub.setEvent('Transfer', Buffer.from(JSON.stringify(transferEvent)));
+
+        return nft;
+    }
+
+    /**
+     * SetTokenURI adds token URI to a given non-fungible token
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} tokenId Unique ID of the non-fungible token to be minted by the msg.sender
+     * @param {String} tokenURI URI containing metadata of the minted non-fungible token
+     * @returns {Boolean} Return whether the burn was successful or not
+    */
+    async SetTokenURI(ctx, tokenId, tokenURI) {
```

Comment:
Should there be an access control check here to ensure that the caller owns the token and is permitted to set the token URI?

Reply:
`SetTokenURI` was removed because we use `MintWithTokenURI` to set a token URI in a single function.